Refactor and cleanup

jnovinger opened the pull request:
Addresses lingering issues from previous PRs as outlined in #4 
- [ ] [Address potential NPE](https://github.com/jnovinger/DemoReader/pull/9#discussion-diff-62273075)

jnovinger committed:
Remove unused import

jnovinger committed:
Refactor POJOs for gson

Simplified by removing apparently unnecessary constructors and
`set<field>()` methods. Took advantage of `@SerializedName` in order to
use the `mField` style guide recommendation and still parse JSON with
different keys.

jnovinger committed:
Refactor `ArticleFeed` to just wrap feed logic/data

jnovinger committed:
Move ArticleFeed out of datamodels package

jnovinger committed:
minor cleanup

jnovinger committed:
Clean up/refactor: layouts, data munging, etc.

With @moagrius's help:
- Remove junk from layouts that wasn't needed, included padding that
  confused the UX.
- Clean up the json parsing, using an extended `ArrayList` to parse a
  top-level json array.

moagrius commented on app/src/main/java/org/novinger/jason/demoreader/ArticleAdapter.java:
```diff
@@ -53,7 +52,7 @@ public ArticleAdapter(ArrayList<Article> articles) {
         Article article = mArticles.get(position);
         holder.mHeadline.setText(article.getHeadline());
-        holder.mByline.setText(article.getPrimaryAuthor().toString());
+        holder.mByline.setText(article.getPrimaryAuthor().getName());
```
any chance of NPE here?  (any chance anything in this chain might be `null`)

jnovinger replied:
Very possibly, will verify.

moagrius commented on app/src/main/java/org/novinger/jason/demoreader/ArticleAdapter.java:
```diff
@@ -53,7 +52,7 @@ public ArticleAdapter(ArrayList<Article> articles) {
-        holder.mByline.setText(article.getPrimaryAuthor().toString());
+        holder.mByline.setText(article.getPrimaryAuthor().getName());
         Glide.with(holder.mContext).load(article.getLeadArtURL()).into(holder.mLeadArt);
     }
```
interesting https://inthecheesefactory.com/blog/get-to-know-glide-recommended-by-google/en

jnovinger replied:
Agreed

moagrius commented on app/src/main/java/org/novinger/jason/demoreader/ArticleListActivity.java:
```diff
@@ -73,30 +59,12 @@ public boolean onOptionsItemSelected(MenuItem item) {
-        if(qty > 0) {
-           return feedUrl + "?quantity=" + Integer.toString(qty);
+        if(mQuantity > 0) {
+           return mFeedUrl + "?quantity=" + Integer.toString(mQuantity);
```
so this is totally fine, and probably how i'd do it, but fwiw there's a lot of old fuddies around that love `String.format`, e.g., `String.format("some string %s", someVariable)`, which has a lot of history, docs, and tricks since it's such a common pattern...

jnovinger replied:
Heh, if I'd known about/remembered `String.format`, I probably would have used it. :P

moagrius commented on app/src/main/java/org/novinger/jason/demoreader/datamodels/Article.java:
```diff
@@ -1,79 +1,65 @@
     public Author getPrimaryAuthor() {
-        if (authors.size() > 0) {
-            return authors.get(0);
+        if (null != mAuthors && mAuthors.size() > 0) {
```
so there are argument to be made for it, but i personally always use the opposite order: `mAuthors != null` - is this a decision, or do you not have an opinion either way?

jnovinger replied:
This is an old habit that came from getting NPE when putting the variable first. I don't remember why, but it was beaten in to me at NISC. I remember you telling me it's not any issue anymore (or maybe never was).

moagrius commented on app/src/main/java/org/novinger/jason/demoreader/datamodels/Article.java:
```diff
@@ -1,79 +1,65 @@
     public String getLeadArtURL() {
-        if (null == lead_art) {
+        if (null == mLeadArt) {
             return "";
```
not sure if you want it or not, but generally a "bad" String is `null`, not empty...

moagrius commented:
+100

jnovinger committed:
Big cleanup/item view refactor

- Create tile effect with `ImageView` and `TextView`s inside a
  `RelativeLayout` per article.
- Embrace R.values.styles
- Remove saved reference to context on individual `ViewHolder`s in favor
  of getting it from one of the constituent views.
- Use a Glid `RequestListener` to put off updating text views until the
  `ImageView` is ready.

jnovinger committed:
Add section label to list item view

Ugh, how did it get to be 1245a already. Damn, this is kind of fun.

At any rate, started also deserializing section name and pub date. Added
section to item view just above headline. Still need to figure out how
to parse the pub date and display it sanely.

jnovinger commented:
Current status:
![selection_540](https://cloud.githubusercontent.com/assets/103499/15065360/81a006fe-1324-11e6-9b79-bc880dbf5cad.png)

moagrius commented:
very nice!
On May 6, 2016 12:50 AM, "Jason Novinger" notifications@github.com wrote:

> Current status:
> [image: selection_540]
> https://cloud.githubusercontent.com/assets/103499/15065360/81a006fe-1324-11e6-9b79-bc880dbf5cad.png
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/jnovinger/DemoReader/pull/9#issuecomment-217353237
